Mathlib/GroupTheory/RegularWreathProduct.lean

```diff
@@ -0,0 +1,301 @@
+    IteratedWreathProduct G (n + 1) = (IteratedWreathProduct G n) ≀ᵣ G := rfl
+
+instance [Finite G] : Finite (IteratedWreathProduct G n) := by
+  induction n with
+  | zero => rw [IteratedWreathProduct_zero]; infer_instance
+  | succ n h => rw [IteratedWreathProduct_succ]; infer_instance
+
+theorem iter_wreath_card [Finite G] : Nat.card (IteratedWreathProduct G n) =
+    Nat.card G ^ (∑ i ∈ Finset.range n, Nat.card G ^ i) := by
+  induction n with
+  | zero => simp
+  | succ n h => rw [IteratedWreathProduct_succ, RegularWreathProduct.card,
+      h, geom_sum_succ, pow_succ, pow_mul']
+
+variable [Group G]
+
+instance : Group (IteratedWreathProduct G n) := by
+ induction n with
+ | zero => rw [IteratedWreathProduct_zero]; infer_instance
+ | succ n ih => rw [IteratedWreathProduct_succ]; infer_instance
+
+lemma mu_eq {p n : ℕ} [hp : Fact (Nat.Prime p)] :
+    (p ^ n).factorial.factorization p = ∑ i ∈ Finset.range n, p ^ i := by
+  rw [← Nat.multiplicity_eq_factorization hp.out (p ^ n).factorial_ne_zero, ← ENat.coe_inj,
+    ← (Nat.finiteMultiplicity_iff.2
+      ⟨hp.out.ne_one, (p ^ n).factorial_pos⟩).emultiplicity_eq_multiplicity]
+  induction n with
+  | zero => simp [hp.out.emultiplicity_one]
+  | succ n h =>
+    rw [pow_succ', hp.out.emultiplicity_factorial_mul, h, Finset.sum_range_succ, ENat.coe_add]
+
+/-- If `α` is equivalent to `β`, then `Perm α` is isomorphic to `Perm β`. -/
+def Equiv.permCongrHom {α β : Type*} (e : α ≃ β) : Equiv.Perm α ≃* Equiv.Perm β where
+  toFun x := e.symm.trans (x.trans e)
+  invFun y := e.trans (y.trans e.symm)
+  left_inv _ := by ext; simp
+  right_inv _ := by ext; simp
+  map_mul' _ _ := by ext; simp
+
+/-- The homomorphism from the wreath product of the Sylow `p`-subgroup and `Z_p` to
+  the subgroup of `Perm (Fin p^(n+1))`. -/
+noncomputable def sylowWreathtoPermHom {p n : ℕ} (D : Sylow p (Equiv.Perm (Fin (p ^ n))))
+    (Z_p : Type*) [Finite Z_p] [Group Z_p] (h : Nat.card Z_p = p) :
+    D ≀ᵣ Z_p →* Equiv.Perm (Fin (p ^ (n + 1))) :=
+  (Equiv.permCongrHom ((Equiv.prodCongrRight fun _ =>
+  (Finite.equivFinOfCardEq h)).trans finProdFinEquiv)).toMonoidHom.comp
+  (RegularWreathProduct.toPerm D Z_p (Fin (p ^ n)))
+
+lemma sylowWreathtoPermHomInj {p n : ℕ} [Fact (Nat.Prime p)]
+    (D : Sylow p (Equiv.Perm (Fin (p ^ n))))
+    (G : Type) [Finite G] [Group G] (h : Nat.card G = p) :
+    Function.Injective (sylowWreathtoPermHom D G h) := by
+  have : Function.Injective (RegularWreathProduct.toPerm D G (Fin (p ^ n))) :=
+    RegularWreathProduct.toPermInj D G (Fin (p ^ n))
+  exact (fun a b => Function.Injective.comp a b)
+    (Equiv.permCongrHom (((Equiv.prodCongrRight fun _ =>
+    (Finite.equivFinOfCardEq h)).trans finProdFinEquiv))).injective this
+
+/-- `Fin (n + 1) → α` and `(Fin n → α) × α` are equivalent. -/
+def Fin.succFunEquiv (α : Type*) (n : ℕ) : (Fin (n + 1) → α) ≃ (Fin n → α) × α :=
+  (Fin.appendEquiv n 1).symm.trans (Equiv.prodCongrRight fun _ ↦ Equiv.funUnique (Fin 1) α)
```

Comment:
Can you move this lemma to `Mathlib.Logic.Equiv.Fin.Basic` (I think that's the best spot).